reworked scenario concept chapter (#31)
mqtt-varroa/varroa-documentation: LaTeX/Parts/Chapter4.tex

%\end{lstlisting}

> \begin{lstlisting}[caption={example XML definition of the Scenario}, captionpos=b, label={lst:scenario}, language=XML]

Reviewer: ```suggestion
\begin{lstlisting}[caption={example for the XML definition of a Scenario}, captionpos=b, label={lst:scenario}, language=XML]
```

Maybe sounds better, rework for all listing captions.

\begin{itemize}
	\item \textbf{id:} the identifier to reference the Subscription.
	\item \textbf{topicGroup:} reference to a Topic Group.
	\item \textbf{wildCard:} When true the topic name pattern of the Topic Group is analysed and all expandable regex are replaced with single level wildcards ("+"). 
	The regex must contain an expandable topic level separator ("/")
\end{itemize}

\section{Wait Patterns} \label{sec:waitPatterns}
\begin{lstlisting}[caption={example XML definition of waitPatterns}, captionpos=b, label={lst:waitPatterns}, language=XML]
<scenario>
	<waitPatterns>
		<waitPattern id="waitPattern-1">
			<subscription>subscription-1</subscription>
			<messagePattern>payload[0-9]{10}</messagePattern>
		</waitPattern>

		<-- definition of more wait patterns -->
	</waitPatterns>
</scenario>
\end{lstlisting}

> A Wait Pattern enables the user to module waiting behaviours where MQTT clients waits for receiving messages in a Subscription (see \ref{sec:subscriptions}) before executing other Commands.

Reviewer: ```suggestion
A Wait Pattern enables the user to model waiting behaviours where MQTT clients wait for incoming messages on a Subscription (see \ref{sec:subscriptions}) before executing other Commands.
```
Not sure about the 'on', but wait was correct, not waits.

As seen in the example above, the following parameters need to be specified:
\begin{itemize}
	\item \textbf{id:} the identifier to reference the Wait Pattern.
	\item \textbf{subscription:} the Subscription the Wait Pattern targets.

> 	\item \textbf{messagePattern (optional):} the message pattern that is waited for in the targeted Subscription.

Reviewer: ```suggestion
	\item \textbf{messagePattern (optional):} the message pattern that is waited for on the targeted Subscription.
```

</scenario>
\end{lstlisting}
A Scenario is divided into one or more Stages.

> These Stages are executed sequentially in the order in which they are specified in the XML document.

Reviewer: ```suggestion
These Stages are executed in the order in which they are specified in the XML document.
```
The sequentially just makes it less readable in my opinion.
The reader can figure that out implicitly.

A Stage contains a number of Lifecycles, which specify the behaviour of Client Groups in the stage.
Lifecycles within a stage are executed in parallel.
To define a stage the user needs to specify:
\begin{itemize}
	\item \textbf{id:} the identifier to reference the Stage.
	\item \textbf{expectedDuration (optional):} the expected amount of time this Action takes to be executed by the whole Client Group. Used for reporting
\end{itemize}

> A Lifecycle contains an amount of Commands (see \ref{sec:commands}) that must be executed by a Client Group.

Reviewer: ```suggestion
A Lifecycle contains an amount of Commands (see \ref{sec:commands}) which are a assigned to be executed by a certain Client Group.
```

To define a Lifecycle the user needs to specify:
\begin{itemize}
	\item \textbf{id:} the identifier to reference the Lifecycle.
	\item \textbf{clientGroupId:} the Client Group that executes the Commands in this Lifecycle.
\end{itemize}

\section{Commands}\label{sec:commands}

> A Command is to be executed by a Client Group, it must be contained in a Lifecycle.

Reviewer: ```suggestion
A Command has to be executed by a Client Group, it must be contained in a Lifecycle.
```

\subsection{connect}
\begin{lstlisting}[caption={example XML definition of a connect command}, captionpos=b, label={lst:connect}, language=XML]
<connect broker="b1" expectedDuration="10s"/>
\end{lstlisting}

> The connect command orders the Client Group to connect to a broker.

Reviewer: ```suggestion
The connect command gives an order to the Client Group to connect to a broker.
```

The Client Group disconnects from the broker it is connected to.

\subsection{publish}
\begin{lstlisting}[caption={example XML definition of a publish commmand}, captionpos=b, label={lst:publish}, language=XML]
<publish topicGroup="tg1" count="10" message="message" rate="100/1s"/>
<publish topicGroup="tg1" count="10" payloadGenerator="randomAlphaNumeric"/>
<publish topicGroup="tg1" message="{{clientId}}" payloadGenerator="mustache"/>
<publish topicGroup="tg1" message="payload[0-9]{10}" payloadGenerator="regex"/>
\end{lstlisting}
A publish Command orders the Client Group to publish messages.
It has the following parameters:
\begin{itemize}
	\item \textbf{topicGroup (required):} references the Topic Group the message is published to (see \ref{sec:topicGroups}). 
	\item \textbf{message (optional):} the argument for the payloadGenerator.
	\item \textbf{payloadGenerator (optional):} the generator for the payload that is sent within the message. The following options exist:
		\begin{itemize}
			\item \textbf{randomAlphaNumeric:} generates random alpha numeric strings as payloads.
			\item \textbf{static (default):} simply uses the message parameter as payload.

> 			\item \textbf{mustache:} generates payloads bases on a mustache.js template given in the message parameter.

Reviewer: ```suggestion
			\item \textbf{mustache:} generates payloads based on a mustache.js template given in the message parameter.
```

\end{itemize}

\subsection{subscribe}
\begin{lstlisting}[caption={example XML definition of a subscibe command}, captionpos=b, label={lst:subscirbe}, language=XML]
<subscribe subscription="sub1" qos="1"/>
\end{lstlisting}

> A subscribe Command orders the Client Group to subscribe.

Reviewer: ```suggestion
A subscribe Command gives an order to the Client Group to subscribe.
```

It has the following parameters:
\begin{itemize}
	\item \textbf{subscription (required):} references the Subscription to be executed (see \ref{sec:subscriptions}).
	\item \textbf{qos (optional):} the quality of service.
	\item \textbf{expectedDuration (optional):} the expected amount of time this action takes to be executed by the whole Client Group. Used for reporting.
\end{itemize}

\subsection{unsubscribe}
\begin{lstlisting}[caption={example XML definition of a unsubscibe command}, captionpos=b, label={lst:unsubscirbe}, language=XML]
<unsubscribe subscription="sub1"/>
\end{lstlisting}

> A unsubscribe Command orders the Client Group to unsubscribe.

Reviewer: ```suggestion
A unsubscribe Command gives and order to the Client Group to unsubscribe.
```

It has the following parameters:
\begin{itemize}
	\item \textbf{subscription (required):} references the Subscription to be terminated (see \ref{sec:subscriptions}).
	\item \textbf{expectedDuration (optional):} the expected amount of time this action takes to be executed by the whole Client Group. Used for reporting.
\end{itemize}

\subsection{wait}
\begin{lstlisting}[caption={example XML definition of a wait command}, captionpos=b, label={lst:wait}, language=XML]
<wait waitPattern="wp1" duration="10s"/>
\end{lstlisting}

> A wait Command orders a Client Group to wait before executing the following Commands.

Reviewer: ```suggestion
A wait Command gives an order to a Client Group to wait before executing the following Commands.
```

It has the following parameter:
\begin{itemize}
	\item \textbf{waitPattern (optional):} references the Wait Pattern (see \ref{sec:waitPatterns}) to waited for.
	\item \textbf{duration (optional):} the amount of time to wait.
	\item \textbf{expectedDuration: (optional):} the expected amount of time this action takes to be executed by the whole Client Group. Used for reporting.
\end{itemize}
If both a Wait Pattern and a duration are given the following Commands are executed after either the duration is exceeded or the Wait Pattern has matched.

\subsection{for}
\begin{lstlisting}[caption={example XML definition of a for command}, captionpos=b, label={lst:for}, language=XML]
<for times="2">
	<connect broker="b1"/>
	<disconnect/>
</for>
\end{lstlisting}

> A for Command orders the Client Group to execute the inner Commands multiple times.

Reviewer: ```suggestion
A for Command gives an order to the Client Group to execute the inner Commands multiple times.
```

It has the following parameters:
\begin{itemize}
	\item \textbf{times:} the amount of times the inner Commands are executed.
	\item \textbf{rate (optional):} the rate at which the inner Commands are executed.
	\item \textbf{expectedDuration: (optional):} the expected amount of time this action takes to be executed by the whole Client Group. Used for reporting.
\end{itemize}

\subsection{rampUp}
\begin{lstlisting}[caption={example XML definition of a rampUp command}, captionpos=b, label={lst:rampUp}, language=XML]
<rampUp duration="10s"/>	
\end{lstlisting}

> The Client Group executes the following Commands of the lifecycle in staggered order.

Reviewer: ```suggestion
The Client Group executes the following Commands of the lifecycle in a staggered manner.
```
